extensions/amp-story/1.0/amp-story-interactive-confetti.js

```diff
@@ -0,0 +1,124 @@
+ */
+
+import {htmlFor} from '../../../src/static-template';
+import {removeChildren} from '../../../src/dom';
+import {setStyles} from '../../../src/style';
+
+/**
+ * Generates the template for the particle wrapper.
+ *
+ * @param {!Element} element
+ * @return {!Element}
+ */
+const buildWrapperTemplate = (element) => {
+  const html = htmlFor(element);
+  return html`
+    <div
+      class="i-amphtml-story-interactive-particle-wrapper"
+      aria-hidden="true"
+    ></div>
+  `;
+};
+
+/**
+ * Generates a template for a particle.
+ *
+ * @param {!Element} element
+ * @return {!Element}
+ */
+const buildParticleTemplate = (element) => {
+  const html = htmlFor(element);
+  return html` <div class="i-amphtml-story-interactive-particle"></div> `;
+};
+
+/**
+ * Creates a radial burst of emoji.
+ * The burst begins from the center of the parent element and
+ * animates to the edges of the parent element.
+ * Nodes are removed from the dom at the end of the animation.
+ *
+ * @param {!Element} rootEl
+ * @param {string} confettiEmoji
+ * @return {void}
+ */
+export function emojiBurst(rootEl, confettiEmoji) {
+  const PARTICLE_COUNT = 5;
+  const SLICE = (Math.PI * 2) / PARTICLE_COUNT;
+
+  const ANGLE_RANDOMNESS = SLICE * 0.2;
+  const FONT_SIZE_RANDOM_RANGE = [30, 50];
+  const ROTATION_RANDOMNESS = 20;
+  const ADDITIONAL_DISTANCE = 5;
+
+  const ANIMATION_IN_DELAY = 300;
+  const ANIMATION_OUT_DELAY = 1000;
+
+  // To calculate particle transform distance.
+  const ROOT_EL_RECT = rootEl./*OK*/ getBoundingClientRect();
+
+  const particleWrapper = buildWrapperTemplate(rootEl);
+  rootEl.appendChild(particleWrapper);
+
+  setTimeout(() => {
+    // Generate particles. Set their ending position, size and rotation.
+    for (let i = 0; i < PARTICLE_COUNT; i++) {
+      const particle = buildParticleTemplate(rootEl);
+      particleWrapper.appendChild(particle);
+      particle.textContent = confettiEmoji;
+
+      const fontSize =
+        randomInRange(FONT_SIZE_RANDOM_RANGE[0], FONT_SIZE_RANDOM_RANGE[1]) +
+        'px';
+      const angle =
+        SLICE * i + randomInRange(-ANGLE_RANDOMNESS, ANGLE_RANDOMNESS);
+      const x =
+        Math.sin(angle) * (ROOT_EL_RECT.width / 2 + ADDITIONAL_DISTANCE);
+      const y =
+        Math.cos(angle) * (ROOT_EL_RECT.height / 2 + ADDITIONAL_DISTANCE);
+      const rotation = randomInRange(-ROTATION_RANDOMNESS, ROTATION_RANDOMNESS);
+      setStyles(particle, {
+        fontSize,
+        transform: `translate(${x}px, ${y}px) rotate(${rotation}deg)`,
+      });
+    }
+    // Scale up particle container.
+    setStyles(particleWrapper, {transform: 'scale(1)'});
+
+    // Animate out the particle wrapper and remove particles from the dom.
+    setTimeout(() => {
+      setStyles(particleWrapper, {
+        transform: `scale(1.2)`,
+        opacity: `0`,
+      });
+
+      particleWrapper.addEventListener('transitionend', () => {
+        removeChildren(particleWrapper);
```

Comment:
Can we recycle the DOM elements so that we don't have to create/delete them every time? Maybe resetting their `transform` and just swapping the `textContent` for each type.

Reply:
Re-using the elements sounds good. Logic would need restructuring to get that woking.

As per our discussion:
- It would be attached to the root `amp-story` element.
- It would be initialized on the `amp-story` level.
- It would listen to `INTERACTIVE_REACT_STATE` to update the emoji and get a reference to the poll id for positioning.

@mszylkowski @gmajoulet what do you think of this approach?

Reply:
Not trying to start a long discussion on GH 😁 but I feel like recycling the elements is premature optimization, unless there are actual performance improvements (that counteract the repositioning / reparenting the element, updating styles to show/hide, always having an extra element, etc). If the performance doesn't change much, the current approach is just simpler (and if we end up having multiple animations, each one wouldn't require a separate recycled element). 

If we recycle the elements, adding it to the story level makes sense but we should contain the code to the amp-story-interactive files if possible (since they will soon be a separate extension). That way you also wouldn't need to subscribe to the REACT_STATE, and you only create the recycled element if the extension was loaded.

Can you add a console.time() and console.timeEnd() and let us know what the timing is for the current approach? [Example here](https://levelup.gitconnected.com/different-ways-to-measure-performance-in-javascript-94785075ab96)

Reply:
Agree with all of this.

.48ms on average to generate particles for the 14 examples we have.

Since we're moving all of this to a separate extension I feel like it's good to keep this decoupled from `amp-story`. 

![generate particles](https://user-images.githubusercontent.com/3860311/87456358-ca89f980-c5d4-11ea-8066-34ab9494802b.png)
![sh](https://user-images.githubusercontent.com/3860311/87456427-ebeae580-c5d4-11ea-94e4-67eee98ec3f1.png)

Reply:
@Enriqe I feel like 0.48ms is not worth trying to optimize in this case, also given the extra code that is needed to make this optimization will have similar performance stats, wdyt?

Reply:
Thanks both for the context & analysis. I don't feel strongly about this so I'll defer to @gmajoulet, just a note that measuring time might not be the most appropriate metric to look at here, especially when using high end laptops and no CPU throttling. Ideally we'd want to test with an average mobile device and see how bad the performance is.

Reply:
Lots of good arguments for both sides, thanks all for driving this conversation!
I agree that re-using DOM elements is great, also agree that premature optimization is bad, also agree that performance should be measured on low end phones.

If the code to re-use the elements was straightforward it'd be worth it, but as @processprocess mentioned the reparenting + moving the logic somewhere else + using the store service makes it a bit complicated for now.
This code execution is not in the way of any critical task like page rendering or page navigation, and we don't really know if publishers will use several quizzes per story. So the current implementation is fine with me :))